torch_xla/csrc/xla_sharding_util.cpp

```diff
@@ -86,6 +86,65 @@ xla::HloModuleProto ShardingUtil::SpmdPartitioningPass(
   return module.get()->ToProto();
 }
 
+std::vector<std::vector<xla::ComputationClient::DataPtr>>
+ShardingUtil::InputHandler(
+    std::vector<xla::ComputationClient::DataPtr> arguments,
+    std::vector<std::string> devices) {
+  std::vector<std::vector<xla::ComputationClient::DataPtr>> arguments_by_device(
+      devices.size(),
+      std::vector<xla::ComputationClient::DataPtr>(arguments.size()));
+  for (int64_t argument_i = 0; argument_i < arguments.size(); ++argument_i) {
+    auto shards =
+        xla::ComputationClient::Get()->GetDataShards(arguments[argument_i]);
+    if (shards.size() > 1) {
+      // Input is sharded across addressable devices
+      for (auto shard : shards) {
+        int64_t device_i = ParseDeviceString(shard->device()).ordinal();
+        arguments_by_device[device_i][argument_i] = shard;
+      }
+    } else {
+      // Input is replicated across addressable devices
+      int64_t source_device_i =
+          ParseDeviceString(shards[0]->device()).ordinal();
+      arguments_by_device[source_device_i][argument_i] = shards[0];
+
+      auto literal = std::make_shared<xla::Literal>(std::move(
+          xla::ComputationClient::Get()->TransferFromServer(shards)[0]));
+      std::vector<xla::ComputationClient::TensorSource> source_tensors;
+      for (int64_t device_i = 0; device_i < devices.size(); ++device_i) {
```

Comment:
This part of the code seems redundant, can we reuse the `CreateTensorsData` logic above. Also please add a comment here explaining why we delay the replicated data uploading until here.